{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d246150e-2694-4ed7-9657-49e26e0ce2c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 1318/1318 [06:49<00:00,  3.22it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Unnamed: 0                                              Title      Year  \\\n",
      "0              0  Caught in the act: revealing the metastatic pr...  ['2013']   \n",
      "1              1  An Integrative Genomics Approach for Associati...  ['2017']   \n",
      "2              2  Integrative Analysis of MicroRNA-Mediated Gene...  ['2017']   \n",
      "3              3  Advanced Development of Primary Pancreatic Org...  ['2018']   \n",
      "4              4  Redox modifications of cysteine residues regul...      2021   \n",
      "...          ...                                                ...       ...   \n",
      "1313        1313  The roles of monkey M1 neuron classes in movem...      2013   \n",
      "1314        1314  Linkage Disequilibrium Mapping of the Replicat...  ['2009']   \n",
      "1315        1315  Tailoring a Combination Preerythrocytic Malari...      2016   \n",
      "1316        1316  Redox Regulatory Mechanism of Transnitrosylati...      2010   \n",
      "1317        1317  Nuclear Organization and Dynamics of 7SK RNA i...  ['2010']   \n",
      "\n",
      "                                           Author names  \\\n",
      "0                  ['Miriam R. Fein', 'Mikala Egeblad']   \n",
      "1     ['Chindo Hicks', 'Ritika Ramani', 'Oliver Sart...   \n",
      "2     ['Ritika Ramani', 'Gail Megason', 'Jason Schal...   \n",
      "3     ['Shurong Hou', 'Hervé Tiriac', 'Banu Priya Sr...   \n",
      "4     ['Huan Yang', 'Peter Lundbäck', 'Lars Ottosson...   \n",
      "...                                                 ...   \n",
      "1313  ['Matthew T. Kaufman', 'Mark M. Churchland', '...   \n",
      "1314  ['Inga Prokopenko', 'Eleftheria Zeggini', 'Rob...   \n",
      "1315  ['Karolis Bauza', 'Erwan Atcheson', 'Tomas Mal...   \n",
      "1316  ['Changgong Wu', 'Tong Liu', 'Wei Chen', 'Shin...   \n",
      "1317  ['Kannanganattu V. Prasanth', 'Matthew Camiolo...   \n",
      "\n",
      "                                        Journal  \\\n",
      "0               ['Disease Models & Mechanisms']   \n",
      "1                        ['Biomarker Insights']   \n",
      "2                        ['Biomarker Insights']   \n",
      "3                            ['Slas Discovery']   \n",
      "4                        ['Molecular Medicine']   \n",
      "...                                         ...   \n",
      "1313             ['Journal of Neurophysiology']   \n",
      "1314                               ['Diabetes']   \n",
      "1315                 ['Infection and Immunity']   \n",
      "1316  ['Molecular & Cellular Proteomics : MCP']   \n",
      "1317          ['Molecular Biology of the Cell']   \n",
      "\n",
      "                                         ORCID      PMID          PMC  \\\n",
      "0                                           []  23616077  ['3634643']   \n",
      "1                                           []  28469398  ['5391982']   \n",
      "2                                           []  28469402  ['5397277']   \n",
      "3                                           []  29673279  ['6013403']   \n",
      "4     ['http://orcid.org/0000-0001-8796-5523']  34098868  ['8185929']   \n",
      "...                                        ...       ...          ...   \n",
      "1313                                        []  23699057  ['3742981']   \n",
      "1314                                        []  19389826  ['2699860']   \n",
      "1315                                        []  26667840  ['4771343']   \n",
      "1316                                        []  20660346  ['2953919']   \n",
      "1317                                        []  20881057  ['2993747']   \n",
      "\n",
      "                             DOI  \\\n",
      "0             10.1242/dmm.009282   \n",
      "1       10.1177/1177271917695810   \n",
      "2       10.1177/1177271917702895   \n",
      "3       10.1177/2472555218766842   \n",
      "4     10.1186/s10020-021-00307-1   \n",
      "...                          ...   \n",
      "1313       10.1152/jn.00892.2011   \n",
      "1314           10.2337/db09-0081   \n",
      "1315        10.1128/IAI.01063-15   \n",
      "1316     10.1074/mcp.M110.000034   \n",
      "1317     10.1091/mbc.E10-02-0105   \n",
      "\n",
      "                                               Abstract all_urls  \\\n",
      "0     ['The prognosis of metastatic cancer in patien...       []   \n",
      "1     ['High-throughput genotyping has enabled disco...       []   \n",
      "2     ['MicroRNAs (miRNAs) regulate the expression o...       []   \n",
      "3     ['Traditional high-throughput drug screening i...       []   \n",
      "4                                                    []       []   \n",
      "...                                                 ...      ...   \n",
      "1313  ['The motor cortices exhibit substantial activ...       []   \n",
      "1314                                                 []       []   \n",
      "1315  ['The leading malaria vaccine candidate, RTS,S...       []   \n",
      "1316  ['Transnitrosylation and denitrosylation are e...       []   \n",
      "1317  ['We have identified 7SK RNA to be enriched in...       []   \n",
      "\n",
      "                                    combined_string_DAS  citation_count  \n",
      "0                                      Not applicable 1            46.0  \n",
      "1                                      Not applicable 2             4.0  \n",
      "2                                      Not applicable 3             6.0  \n",
      "3                                      Not applicable 4            93.0  \n",
      "4     All\\ndata supporting the findings of this stud...            22.0  \n",
      "...                                                 ...             ...  \n",
      "1313                                Not applicable 1314            53.0  \n",
      "1314                                Not applicable 1315            23.0  \n",
      "1315                                Not applicable 1316            17.0  \n",
      "1316                                Not applicable 1317           106.0  \n",
      "1317                                Not applicable 1318            48.0  \n",
      "\n",
      "[1318 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "#test jupyternotebook- now all functions EXCEPT classify_das.py can be run as individual functions\n",
    "from DAS_extract_V2 import grab_PMIDs_from_csv\n",
    "from DAS_extract_V2 import isolate_rows_by_PMID\n",
    "from DAS_extract_V2 import extract_files\n",
    "from DAS_extract_V2 import get_nxml\n",
    "from DAS_extract_V2 import citation_count\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#a test to see if it works in jupyter notebook\n",
    "file = \"example_files/PMC_parser_output.csv\"\n",
    "df1 = pd.read_csv(file)\n",
    "df2 = df1['DOI']\n",
    "df1['citation_count'] = citation_count(df2)\n",
    "print(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede3a559-f493-4ef4-b3a9-e363f0676323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %run /path/to/my_script.py FOR CLASSIFY DAS python file \n",
    "# !python /path/to/my_script.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c808ad76-8193-4e0f-bc13-e1be43362954",
   "metadata": {},
   "outputs": [],
   "source": [
    "#STEP 1 -grab PMIDs from CSHL digital repo CSV \n",
    "from DAS_extract_V2 import grab_PMIDs_from_csv\n",
    "from DAS_extract_V2 import isolate_rows_by_PMID\n",
    "from DAS_extract_V2 import extract_files\n",
    "from DAS_extract_V2 import get_nxml\n",
    "\n",
    "datafile = \"example_files/2023_07_06_CSHL_articles_2007-2022_from_IR.csv\"\n",
    "pmids_output_file = \"example_files/pmids_07_22.txt\"\n",
    "\n",
    "grab_PMIDs_from_csv(datafile,pmids_output_file)\n",
    "\n",
    "#STEP2 check OA database CSV for files to download-oa_file_list was not able to be uploaded to Github, you must download your own copy from\n",
    "#go to the Base FTP URL https://ftp.ncbi.nlm.nih.gov/pub/pmc/\n",
    "isolate_rows_by_PMID('example_files/pmids_07_22.txt', 'example_files/oa_file_list.csv', 'example_files/output1_07_22_PMC.csv')\n",
    "\n",
    "#STEP 3 - using the links in the output file and download zipped PMC folders containing .nxml, .pdf, figures and data into a directory (download_PMC fxn)\n",
    "#STEP 3A- Each downladed file will look like PMC{ID Number}.tar.gz, using extract_files function to unzip them \n",
    "\n",
    "download_PMC('example_files/output1_07_22_PMC.csv', \"PMC_folder_07_22\")\n",
    "folder = 'example_files/PMC_folder_07_22'\n",
    "extract_files(folder)\n",
    "\n",
    "#STEP 4 grab the full text XML from each PMC folder and create a folder with all articles\n",
    "source_folder = 'example_files/PMC_folder_07_22'\n",
    "target_folder = 'example_files/final_xml_07_22'\n",
    "file_extension = '.nxml'\n",
    "\n",
    "get_nxml(source_folder,target_folder,file_extension)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27dbbce1-7ba0-45ff-af18-546a7619a536",
   "metadata": {},
   "outputs": [],
   "source": [
    "from DAS_extract_V2 import parse_xml_directory\n",
    "#IN THIS STEP WE ARE PARSING THROUGH EACH FULL TEXT XML and returning key article metadata including DAS and url links\n",
    "\n",
    "xml_directory = '/Users/muthuku/Desktop/final_xml_07_22'\n",
    "df1 = parse_xml_directory(xml_directory)\n",
    "#DATAFRAME cleanup steps - run these after the main function above\n",
    "# create a combined DAS column to include results from all methods of DAS collection \n",
    "df1['combined_DAS'] = df1['Data Availability 1'] + df1['Data Availability 2'] + df1['Data Availability 3'] + df1['Data Availability 4'] + df1['Data Availability 5'] \n",
    "df1['unique_combined_das'] = df1['combined_DAS'].apply(lambda x: list(set(x)))\n",
    "count = 0\n",
    "for n in df1['unique_combined_das']:\n",
    "\tcount += 1\n",
    "\tif n == []:\n",
    "\t\tn.append(\"Not applicable \" + str(count))\n",
    "\n",
    "#combine URLs as well \n",
    "df1['all_urls'] = df1['URL1'] +df1['URL2'] + df1['URL3'] +df1['URL4'] + df1['URL5']\n",
    "#drop individual columns \n",
    "column_list = ['Data Availability 1','Data Availability 2','Data Availability 3','Data Availability 4','Data Availability 5','URL1', 'URL2','URL3','URL4','URL5', 'combined_DAS']\n",
    "df2= df1.drop(columns = column_list)\n",
    "#convert the list of seperate strings into a single string- join all sentences together and remove commas which interfere with ML \n",
    "df2['combined_string_DAS'] = df2['unique_combined_das'].apply(lambda x: ' '.join(x))\n",
    "df2['combined_string_DAS'] = df2['combined_string_DAS'].str.replace(',', '')\n",
    "df2 = df2.drop(['unique_combined_das'], axis = 1)\n",
    "df2.to_csv(\"PMC_parser_output.csv\", encoding = \"utf-8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bde77e3-25d7-46d9-94e3-e8d5d135685f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from DAS_extract_V2 import citation_count\n",
    "#STEP3 -citation counts for each article using SCOPUS API and using article DOIs as an input \n",
    "file = \"example_files/PMC_parser_output.csv\"\n",
    "df1 = pd.read_csv(file)\n",
    "df2 = df1['DOI']\n",
    "df1['citation_count'] = citation_count(df2)\n",
    "print(df1)\n",
    "\n",
    "df1.to_csv('og_df_with_citations1.csv', encoding = \"utf-8\", index = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2e888bd-be5a-4915-bd25-5c61d85d213f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%run ./DAS_extract_V2/classify_das2.py \n",
    "#!python3 ./DAS_extract_V2/classify_das2.py\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7077831f-73cb-47bb-9545-65865965b849",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
